nexus/src/app/sagas/region_replacement_finish.rs

```diff
@@ -0,0 +1,345 @@
+
+    Ok(())
+}
+
+async fn srrf_update_request_record(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let params = sagactx.saga_params::<Params>()?;
+    let osagactx = sagactx.user_data();
+    let datastore = osagactx.datastore();
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let saga_id = sagactx.lookup::<Uuid>("saga_id")?;
+
+    // Now that the region has been deleted, update the replacement request
+    // record to 'Complete' and clear the operating saga id. There is no undo
+    // step for this, it should succeed idempotently.
+    datastore
+        .set_region_replacement_complete(&opctx, params.request.id, saga_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+pub(crate) mod test {
+    use crate::{
+        app::saga::create_saga_dag,
+        app::sagas::region_replacement_finish::Params,
+        app::sagas::region_replacement_finish::SagaRegionReplacementFinish,
+    };
+    use async_bb8_diesel::AsyncRunQueryDsl;
+    use chrono::Utc;
+    use nexus_db_model::Region;
+    use nexus_db_model::RegionReplacement;
+    use nexus_db_model::RegionReplacementState;
+    use nexus_db_model::Volume;
+    use nexus_db_queries::authn::saga::Serialized;
+    use nexus_db_queries::context::OpContext;
+    use nexus_test_utils_macros::nexus_test;
+    use sled_agent_client::types::CrucibleOpts;
+    use sled_agent_client::types::VolumeConstructionRequest;
+    use uuid::Uuid;
+
+    type ControlPlaneTestContext =
+        nexus_test_utils::ControlPlaneTestContext<crate::Server>;
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_region_replacement_finish_saga(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let nexus = &cptestctx.server.server_context().nexus;
+        let datastore = nexus.datastore();
+        let opctx = OpContext::for_tests(
+            cptestctx.logctx.log.clone(),
+            datastore.clone(),
+        );
+
+        // Manually insert required records
+        let old_region_volume_id = Uuid::new_v4();
+        let new_volume_id = Uuid::new_v4();
+
+        let replaced_region = {
+            let dataset_id = Uuid::new_v4();
+            Region::new(
+                dataset_id,
+                old_region_volume_id,
+                512_i64.try_into().unwrap(),
+                10,
+                10,
+            )
+        };
+
+        {
+            let conn = datastore.pool_connection_for_tests().await.unwrap();
+
+            use nexus_db_model::schema::region::dsl;
+            diesel::insert_into(dsl::region)
+                .values(replaced_region.clone())
+                .execute_async(&*conn)
+                .await
+                .unwrap();
+        }
+
+        let volume_construction_request = VolumeConstructionRequest::Volume {
+            id: old_region_volume_id,
+            block_size: 0,
```

Comment:
Should be a pretty quick replacement with that block size...

Reply:
:speedboat: